src/dparse/strings.d

```diff
@@ -0,0 +1,694 @@
+	assert(unescapeString!(InvalidEscapeAction.keep)(`q"/xd"`) == "/xd");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`q"/xd"`) == "xd");
+
+	assertThrown!ConvException(unescapeString(`"\x"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\x"`) == "\\x");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\x"`) == "");
+
+	assertThrown!ConvException(unescapeString(`"\u0"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\u0"`) == "\\u0");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\u0"`) == "");
+
+	assertThrown!ConvException(unescapeString(`"\U0000000"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\U0000000"`) == "\\U0000000");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\U0000000"`) == "");
+
+	assertThrown!ConvException(unescapeString(`"\xAG"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\xAG"`) == "\\xAG");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\xAG"`) == "");
+
+	assertThrown!ConvException(unescapeString(`"\u00AG"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\u00AG"`) == "\\u00AG");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\u00AG"`) == "");
+
+	assertThrown!ConvException(unescapeDoubleQuotedContent(`a\`));
+	assert(unescapeDoubleQuotedContent!(InvalidEscapeAction.keep)(`a\`) == "a\\");
+	assert(unescapeDoubleQuotedContent!(InvalidEscapeAction.skip)(`a\`) == "a");
+
+	assertThrown!ConvException(unescapeString(`"\z"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`"\z"`) == "\\z");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`"\z"`) == "z");
+
+	assert(parseHexStringContent("") == "");
+
+	assertThrown!ConvException(unescapeString(`x"AG"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`x"AG"`) == "AG");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`x"AG"`) == "");
+
+	assertThrown!ConvException(unescapeString(`x"A"`));
+	assert(unescapeString!(InvalidEscapeAction.keep)(`x"A"`) == "A");
+	assert(unescapeString!(InvalidEscapeAction.skip)(`x"A"`) == "");
+}
+
+private string unescapeDoubleQuotedContent(
+	InvalidEscapeAction invalidEscapeAction = InvalidEscapeAction.error
+)(
+	string input
+)
+{
+	auto escape = input.indexOf('\\');
+	if (escape == -1)
+		return input;
+
+	auto ret = appender!string;
+	ret.reserve(input.length);
+	size_t start = 0;
+
+	bool requireMinLength(size_t length)
+	{
+		if (escape + length >= input.length)
+		{
+			final switch (invalidEscapeAction)
+			{
+			case InvalidEscapeAction.keep:
+				ret ~= input[start .. $];
+				start = input.length;
+				return false;
+			case InvalidEscapeAction.skip:
+				start = input.length;
+				return false;
+			case InvalidEscapeAction.error:
+				throw new ConvException("Unfinished escape at end of string");
+			}
+		}
+		else
+		{
+			return true;
+		}
+	}
+
+	void errorInvalidCharacter(size_t continueAt)
+	{
+		final switch (invalidEscapeAction)
+		{
+		case InvalidEscapeAction.keep:
+			ret ~= input[start .. start = continueAt];
```

Comment:
Can this assignment be moved somewhere else? I'm guessing that it evaluates the left of the `..` first, then does the assignment, and then evaluates the `start` on the right side, but that's not easy to read.